text/0000-bench-utils.md

```diff
@@ -0,0 +1,210 @@
+having any effect (other than inhibiting optimizations).
+
+For example ([`rust.godbolt.org`](https://godbolt.org/g/YP2GCJ)):
+
+```rust
+fn foo(x: i32) -> i32{ 
+  hint::black_box(2 + x);
+  3
+}
+let a = foo(2);
+```
+
+In the call to `foo(2)` the compiler is allowed to simplify the expression `2 +
+x` down to `4`. However, even though `4` is not read by anything afterwards, it
+must be computed and materialized, for example, by storing it into memory, a
+register, etc. because `black_box` could try to read it.
+
+### Benchmarking `Vec::push`
+
+The `hint::black_box` is useful for producing synthetic benchmarks that more
+accurately represent the behavior of a real application. In the following
+snippet, the function `bench` executes `Vec::push` 4 times in a loop:
+
+```rust
+fn push_cap(v: &mut Vec<i32>) {
+    for i in 0..4 {
+      v.push(i);
+    }
+}
+
+pub fn bench_push() -> Duration { 
+    let mut v = Vec::with_capacity(4);
+    let now = Instant::now();
+    push_cap(&mut v);
+    now.elapsed()
+}
+```
+
+Here, we allocate the `Vec`, push into it without growing its capacity, and drop
+it, without ever using it for anything. If we look at the assembly
+(https://rust.godbolt.org/z/wDckJF):
+
+
+```asm
+example::bench_push:
+  sub rsp, 24
+  call std::time::Instant::now@PLT
+  mov qword ptr [rsp + 8], rax
+  mov qword ptr [rsp + 16], rdx
+  lea rdi, [rsp + 8]
+  call std::time::Instant::elapsed@PLT
+  add rsp, 24
+  ret
+```
+
+it is pretty amazing: LLVM has actually managed to completely optimize the `Vec`
+allocation and call to `push_cap` away! In our real application, we would
+probably use the vector for something, preventing all of these optimizations
+from triggering, but in this synthetic benchmark LLVM optimizations are
+producing a benchmark that won't tell us anything about the cost of `Vec::push`.
+
+We can use hint::black_box to create a more realistic synthetic benchmark. The
+compiler has to assume that `black_box` observes and mutates its argument, hence
+it cannot optimize away the whole benchmark (https://rust.godbolt.org/z/CeXmxN):
+
+```rust
+fn push_cap(v: &mut Vec<i32>) {
+    for i in 0..4 {
+        black_box(v.as_ptr());
+        v.push(black_box(i));
+        black_box(v.as_ptr());
+    }
+}
+```
+
+that prevents LLVM from assuming anything about the vector across the calls to
+`Vec::push`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The 
+
+```
+mod core::hint {
+    /// An _unknown_ unsafe function that returns `x`.
+    pub fn black_box<T>(x: T) -> T;
+}
+```
+
+is an _unknown_ function that can perform any valid operation on `x` that Rust
+is allowed to perform without introducing undefined behavior in the calling
+code. You can rely on `black_box` being a `NOP` just returning `x`, but the
+compiler will optimize under the pessimistic assumption that `black_box` might
+do anything with the data it got.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+TBD.
```

Comment:
This should be filled in before we can move forward. (thanks @Centril for pointing this out)